File: backend/src/engine/arena.py
"""Bot-vs-Bot arena — run N games between strategies and report results."""

from __future__ import annotations

import math
import time
from dataclasses import dataclass, field
from typing import Callable

from src.engine.bot_strategy import BotStrategy
from src.engine.game_simulator import (
    SimulationState,
    apply_action_and_resolve,
)
from src.engine.models import Action, GameConfig, GameResult, Player, PlayerId
from src.engine.protocol import GamePlugin


@dataclass
class ArenaResult:
    """Aggregated results from an arena run."""

    num_games: int
    wins: dict[str, int]
    draws: int
    total_scores: dict[str, list[float]]
    game_durations_ms: list[float]

    def win_rate(self, name: str) -> float:
        return self.wins.get(name, 0) / max(self.num_games, 1)

    def avg_score(self, name: str) -> float:
        scores = self.total_scores.get(name, [])
        return sum(scores) / max(len(scores), 1)

    def score_stddev(self, name: str) -> float:
        scores = self.total_scores.get(name, [])
        if len(scores) < 2:
            return 0.0
        avg = self.avg_score(name)
        variance = sum((s - avg) ** 2 for s in scores) / (len(scores) - 1)
        return math.sqrt(variance)

    def confidence_interval_95(self, name: str) -> tuple[float, float]:
        """95% Wilson score confidence interval for win rate."""
        n = self.num_games
        if n == 0:
            return (0.0, 0.0)
        p = self.win_rate(name)
        z = 1.96
        denom = 1 + z**2 / n
        center = (p + z**2 / (2 * n)) / denom
        margin = z * math.sqrt((p * (1 - p) + z**2 / (4 * n)) / n) / denom
        return (max(0.0, center - margin), min(1.0, center + margin))

    def summary(self) -> str:
        lines = [f"Arena Results ({self.num_games} games)"]
        lines.append("=" * 60)
        for name in self.wins:
            wr = self.win_rate(name)
            ci_lo, ci_hi = self.confidence_interval_95(name)
            avg = self.avg_score(name)
            std = self.score_stddev(name)
            lines.append(
                f"  {name:>12s}: {self.wins[name]:3d} wins "
                f"({wr:5.1%})  "
                f"[95% CI: {ci_lo:.1%}-{ci_hi:.1%}]  "
                f"avg={avg:5.1f} +/- {std:4.1f}"
            )
        lines.append(f"  {'Draws':>12s}: {self.draws}")
        if self.game_durations_ms:
            avg_ms = sum(self.game_durations_ms) / len(self.game_durations_ms)
            total_s = sum(self.game_durations_ms) / 1000
            lines.append(f"  Avg game: {avg_ms:.0f}ms  |  Total: {total_s:.1f}s")
        return "\n".join(lines)


def run_arena(
    plugin: GamePlugin,
    strategies: dict[str, BotStrategy],
    num_games: int = 100,
    base_seed: int = 0,
    num_players: int = 2,
    game_options: dict | None = None,
    alternate_seats: bool = True,
    progress_callback: Callable[[int, int], None] | None = None,
) -> ArenaResult:
    """Run *num_games* between the given strategies and return aggregated stats.

    Parameters
    ----------
    plugin:
        The game plugin to use.
    strategies:
        Mapping of ``strategy_name -> BotStrategy``.  Must have exactly
        *num_players* entries.
    num_games:
        How many games to play.
    base_seed:
        Game *i* uses ``random_seed = base_seed + i``.
    alternate_seats:
        Rotate seat assignments each game so every strategy plays
        each seat equally.
    progress_callback:
        Called with ``(games_completed, total_games)`` after each game.
    """
    strategy_names = list(strategies.keys())
    assert len(strategy_names) == num_players, (
        f"Need exactly {num_players} strategies, got {len(strategy_names)}"
    )

    result = ArenaResult(
        num_games=num_games,
        wins={n: 0 for n in strategy_names},
        draws=0,
        total_scores={n: [] for n in strategy_names},
        game_durations_ms=[],
    )

    for game_idx in range(num_games):
        seed = base_seed + game_idx

        # Determine seat assignment
        if alternate_seats:
            seat_assignment = [
                strategy_names[(i + game_idx) % num_players]
                for i in range(num_players)
            ]
        else:
            seat_assignment = strategy_names[:num_players]

        players = [
            Player(
                player_id=PlayerId(f"p{i}"),
                display_name=seat_assignment[i],
                seat_index=i,
                is_bot=True,
                bot_id=seat_assignment[i],
            )
            for i in range(num_players)
        ]

        pid_to_strategy = {
            f"p{i}": strategies[seat_assignment[i]]
            for i in range(num_players)
        }
        pid_to_name = {f"p{i}": seat_assignment[i] for i in range(num_players)}

        config = GameConfig(random_seed=seed, options=game_options or {})

        t0 = time.monotonic()
        game_result = _play_one_game(plugin, players, config, pid_to_strategy)
        elapsed_ms = (time.monotonic() - t0) * 1000
        result.game_durations_ms.append(elapsed_ms)

        if game_result is None:
            result.draws += 1
            for name in strategy_names:
                result.total_scores[name].append(0.0)
        else:
            for pid, score in game_result.final_scores.items():
                name = pid_to_name.get(pid)
                if name:
                    result.total_scores[name].append(score)

            if len(game_result.winners) == 1:
                winner_name = pid_to_name.get(game_result.winners[0])
                if winner_name:
                    result.wins[winner_name] += 1
            else:
                result.draws += 1

        if progress_callback:
            progress_callback(game_idx + 1, num_games)

    return result


def _play_one_game(
    plugin: GamePlugin,
    players: list[Player],
    config: GameConfig,
    pid_to_strategy: dict[str, BotStrategy],
) -> GameResult | None:
    """Play a single game synchronously. Returns GameResult or None."""
    game_data, phase, _ = plugin.create_initial_state(players, config)

    state = SimulationState(
        game_data=game_data,
        phase=phase,
        players=players,
        scores={p.player_id: 0.0 for p in players},
    )

    # Resolve initial auto-resolve phases
    _resolve_auto(plugin, state)

    max_iterations = 500
    for _ in range(max_iterations):
        if state.game_over is not None:
            break

        if state.phase.auto_resolve:
            _resolve_auto(plugin, state)
            continue

        acting_pid = _get_acting_pid(state.phase)
        if acting_pid is None:
            break

        strategy = pid_to_strategy.get(acting_pid)
        if strategy is None:
            break

        chosen = strategy.choose_action(
            state.game_data, state.phase, PlayerId(acting_pid), plugin
        )

        action_type = state.phase.expected_actions[0].action_type
        action = Action(
            action_type=action_type,
            player_id=PlayerId(acting_pid),
            payload=chosen,
        )
        apply_action_and_resolve(plugin, state, action)

    return state.game_over


def _resolve_auto(plugin: GamePlugin, state: SimulationState) -> None:
    """Resolve auto-resolve phases until a player-action phase or game over."""
    max_auto = 50
    while state.phase.auto_resolve and not state.game_over and max_auto > 0:
        max_auto -= 1
        pid = PlayerId("system")
        pi = state.phase.metadata.get("player_index")
        if pi is not None and pi < len(state.players):
            pid = state.players[pi].player_id
        synthetic = Action(action_type=state.phase.name, player_id=pid)
        apply_action_and_resolve(plugin, state, synthetic)


def _get_acting_pid(phase) -> str | None:
    if phase.expected_actions:
        return phase.expected_actions[0].player_id
    return None

File: game-engine/src/engine/arena.rs
//! Bot-vs-bot arena runner.
//! Mirrors backend/src/engine/arena.py.

use std::collections::HashMap;
use std::time::Instant;

use crate::engine::bot_strategy::{BotStrategy, TypedBotStrategy};
use crate::engine::models::*;
use crate::engine::plugin::{GamePlugin, TypedGamePlugin};
use crate::engine::simulator::{
    apply_action_and_resolve, apply_action_and_resolve_typed,
    SimulationState, TypedSimulationState,
};

/// Aggregated results from an arena run.
pub struct ArenaResult {
    pub num_games: usize,
    pub wins: HashMap<String, usize>,
    pub draws: usize,
    pub total_scores: HashMap<String, Vec<f64>>,
    pub game_durations_ms: Vec<f64>,
}

impl ArenaResult {
    pub fn win_rate(&self, name: &str) -> f64 {
        *self.wins.get(name).unwrap_or(&0) as f64 / self.num_games.max(1) as f64
    }

    pub fn avg_score(&self, name: &str) -> f64 {
        let scores = self.total_scores.get(name);
        match scores {
            Some(s) if !s.is_empty() => s.iter().sum::<f64>() / s.len() as f64,
            _ => 0.0,
        }
    }

    pub fn score_stddev(&self, name: &str) -> f64 {
        let scores = match self.total_scores.get(name) {
            Some(s) if s.len() >= 2 => s,
            _ => return 0.0,
        };
        let avg = self.avg_score(name);
        let variance = scores.iter().map(|s| (s - avg).powi(2)).sum::<f64>() / (scores.len() - 1) as f64;
        variance.sqrt()
    }

    pub fn confidence_interval_95(&self, name: &str) -> (f64, f64) {
        let n = self.num_games;
        if n == 0 {
            return (0.0, 0.0);
        }
        let p = self.win_rate(name);
        let z = 1.96_f64;
        let denom = 1.0 + z * z / n as f64;
        let center = (p + z * z / (2.0 * n as f64)) / denom;
        let margin = z * ((p * (1.0 - p) + z * z / (4.0 * n as f64)) / n as f64).sqrt() / denom;
        ((center - margin).max(0.0), (center + margin).min(1.0))
    }

    pub fn summary(&self) -> String {
        let mut lines = vec![format!("Arena Results ({} games)", self.num_games)];
        lines.push("=".repeat(60));
        for name in self.wins.keys() {
            let wr = self.win_rate(name);
            let (ci_lo, ci_hi) = self.confidence_interval_95(name);
            let avg = self.avg_score(name);
            let std = self.score_stddev(name);
            lines.push(format!(
                "  {:>12}: {:3} wins ({:5.1}%)  [95% CI: {:.1}%-{:.1}%]  avg={:5.1} +/- {:4.1}",
                name,
                self.wins[name],
                wr * 100.0,
                ci_lo * 100.0,
                ci_hi * 100.0,
                avg,
                std,
            ));
        }
        lines.push(format!("  {:>12}: {}", "Draws", self.draws));
        if !self.game_durations_ms.is_empty() {
            let avg_ms = self.game_durations_ms.iter().sum::<f64>() / self.game_durations_ms.len() as f64;
            let total_s = self.game_durations_ms.iter().sum::<f64>() / 1000.0;
            lines.push(format!("  Avg game: {:.0}ms  |  Total: {:.1}s", avg_ms, total_s));
        }
        lines.join("\n")
    }
}

/// Run `num_games` between the given strategies and return aggregated stats.
pub fn run_arena(
    plugin: &dyn GamePlugin,
    strategies: &HashMap<String, Box<dyn BotStrategy>>,
    num_games: usize,
    base_seed: u64,
    num_players: usize,
    game_options: Option<serde_json::Value>,
    alternate_seats: bool,
    progress_callback: Option<&dyn Fn(usize, usize)>,
) -> ArenaResult {
    let strategy_names: Vec<String> = strategies.keys().cloned().collect();
    assert_eq!(strategy_names.len(), num_players);

    let mut result = ArenaResult {
        num_games,
        wins: strategy_names.iter().map(|n| (n.clone(), 0)).collect(),
        draws: 0,
        total_scores: strategy_names.iter().map(|n| (n.clone(), Vec::new())).collect(),
        game_durations_ms: Vec::new(),
    };

    for game_idx in 0..num_games {
        let seed = base_seed + game_idx as u64;

        let seat_assignment: Vec<String> = if alternate_seats {
            (0..num_players)
                .map(|i| strategy_names[(i + game_idx) % num_players].clone())
                .collect()
        } else {
            strategy_names[..num_players].to_vec()
        };

        let players: Vec<Player> = (0..num_players)
            .map(|i| Player {
                player_id: format!("p{}", i),
                display_name: seat_assignment[i].clone(),
                seat_index: i as i32,
                is_bot: true,
                bot_id: Some(seat_assignment[i].clone()),
            })
            .collect();

        let pid_to_strategy: HashMap<String, &dyn BotStrategy> = (0..num_players)
            .map(|i| (format!("p{}", i), strategies[&seat_assignment[i]].as_ref()))
            .collect();

        let pid_to_name: HashMap<String, String> = (0..num_players)
            .map(|i| (format!("p{}", i), seat_assignment[i].clone()))
            .collect();

        let config = GameConfig {
            random_seed: Some(seed),
            options: game_options.clone().unwrap_or(serde_json::json!({})),
        };

        let t0 = Instant::now();
        let game_result = play_one_game(plugin, &players, &config, &pid_to_strategy);
        let elapsed_ms = t0.elapsed().as_secs_f64() * 1000.0;
        result.game_durations_ms.push(elapsed_ms);

        match game_result {
            None => {
                result.draws += 1;
                for name in &strategy_names {
                    result.total_scores.get_mut(name).unwrap().push(0.0);
                }
            }
            Some(gr) => {
                for (pid, score) in &gr.final_scores {
                    if let Some(name) = pid_to_name.get(pid) {
                        result.total_scores.get_mut(name).unwrap().push(*score);
                    }
                }

                if gr.winners.len() == 1 {
                    if let Some(name) = pid_to_name.get(&gr.winners[0]) {
                        *result.wins.get_mut(name).unwrap() += 1;
                    }
                } else {
                    result.draws += 1;
                }
            }
        }

        if let Some(cb) = progress_callback {
            cb(game_idx + 1, num_games);
        }
    }

    result
}

fn play_one_game(
    plugin: &dyn GamePlugin,
    players: &[Player],
    config: &GameConfig,
    pid_to_strategy: &HashMap<String, &dyn BotStrategy>,
) -> Option<GameResult> {
    let (game_data, phase, _) = plugin.create_initial_state(players, config);

    let mut state = SimulationState {
        game_data,
        phase,
        players: players.to_vec(),
        scores: players.iter().map(|p| (p.player_id.clone(), 0.0)).collect(),
        game_over: None,
    };

    // Resolve initial auto-resolve phases
    resolve_auto(plugin, &mut state);

    let max_iterations = 500;
    for _ in 0..max_iterations {
        if state.game_over.is_some() {
            break;
        }

        if state.phase.auto_resolve {
            resolve_auto(plugin, &mut state);
            continue;
        }

        let acting_pid = if !state.phase.expected_actions.is_empty() {
            state.phase.expected_actions[0].player_id.clone()
        } else {
            break;
        };

        let strategy = match pid_to_strategy.get(&acting_pid) {
            Some(s) => *s,
            None => break,
        };

        let chosen = strategy.choose_action(
            &state.game_data,
            &state.phase,
            &acting_pid,
            plugin,
            players,
        );

        let action_type = state.phase.expected_actions[0].action_type.clone();
        let action = Action {
            action_type,
            player_id: acting_pid,
            payload: chosen,
        };
        apply_action_and_resolve(plugin, &mut state, &action);
    }

    state.game_over
}

// ================================================================== //
//  Typed arena — zero-JSON hot path
// ================================================================== //

/// Run typed arena: `num_games` between typed strategies.
pub fn run_arena_typed<P: TypedGamePlugin>(
    plugin: &P,
    strategies: &HashMap<String, Box<dyn TypedBotStrategy<P>>>,
    num_games: usize,
    base_seed: u64,
    num_players: usize,
    game_options: Option<serde_json::Value>,
    alternate_seats: bool,
    progress_callback: Option<&dyn Fn(usize, usize)>,
) -> ArenaResult {
    let strategy_names: Vec<String> = strategies.keys().cloned().collect();
    assert_eq!(strategy_names.len(), num_players);

    let mut result = ArenaResult {
        num_games,
        wins: strategy_names.iter().map(|n| (n.clone(), 0)).collect(),
        draws: 0,
        total_scores: strategy_names.iter().map(|n| (n.clone(), Vec::new())).collect(),
        game_durations_ms: Vec::new(),
    };

    for game_idx in 0..num_games {
        let seed = base_seed + game_idx as u64;

        let seat_assignment: Vec<String> = if alternate_seats {
            (0..num_players)
                .map(|i| strategy_names[(i + game_idx) % num_players].clone())
                .collect()
        } else {
            strategy_names[..num_players].to_vec()
        };

        let players: Vec<Player> = (0..num_players)
            .map(|i| Player {
                player_id: format!("p{}", i),
                display_name: seat_assignment[i].clone(),
                seat_index: i as i32,
                is_bot: true,
                bot_id: Some(seat_assignment[i].clone()),
            })
            .collect();

        let pid_to_strategy: HashMap<String, &dyn TypedBotStrategy<P>> = (0..num_players)
            .map(|i| (format!("p{}", i), strategies[&seat_assignment[i]].as_ref()))
            .collect();

        let pid_to_name: HashMap<String, String> = (0..num_players)
            .map(|i| (format!("p{}", i), seat_assignment[i].clone()))
            .collect();

        let config = GameConfig {
            random_seed: Some(seed),
            options: game_options.clone().unwrap_or(serde_json::json!({})),
        };

        let t0 = Instant::now();
        let game_result = play_one_game_typed(plugin, &players, &config, &pid_to_strategy);
        let elapsed_ms = t0.elapsed().as_secs_f64() * 1000.0;
        result.game_durations_ms.push(elapsed_ms);

        match game_result {
            None => {
                result.draws += 1;
                for name in &strategy_names {
                    result.total_scores.get_mut(name).unwrap().push(0.0);
                }
            }
            Some(gr) => {
                for (pid, score) in &gr.final_scores {
                    if let Some(name) = pid_to_name.get(pid) {
                        result.total_scores.get_mut(name).unwrap().push(*score);
                    }
                }

                if gr.winners.len() == 1 {
                    if let Some(name) = pid_to_name.get(&gr.winners[0]) {
                        *result.wins.get_mut(name).unwrap() += 1;
                    }
                } else {
                    result.draws += 1;
                }
            }
        }

        if let Some(cb) = progress_callback {
            cb(game_idx + 1, num_games);
        }
    }

    result
}

fn play_one_game_typed<P: TypedGamePlugin>(
    plugin: &P,
    players: &[Player],
    config: &GameConfig,
    pid_to_strategy: &HashMap<String, &dyn TypedBotStrategy<P>>,
) -> Option<GameResult> {
    // Create initial state via JSON interface, then decode to typed
    let (game_data, phase, _) = plugin.create_initial_state(players, config);
    let typed_state = plugin.decode_state(&game_data);

    let mut state = TypedSimulationState {
        state: typed_state,
        phase,
        players: players.to_vec(),
        scores: players.iter().map(|p| (p.player_id.clone(), 0.0)).collect(),
        game_over: None,
    };

    // Resolve initial auto-resolve phases
    resolve_auto_typed(plugin, &mut state);

    let max_iterations = 500;
    for _ in 0..max_iterations {
        if state.game_over.is_some() {
            break;
        }

        if state.phase.auto_resolve {
            resolve_auto_typed(plugin, &mut state);
            continue;
        }

        let acting_pid = if !state.phase.expected_actions.is_empty() {
            state.phase.expected_actions[0].player_id.clone()
        } else {
            break;
        };

        let strategy = match pid_to_strategy.get(&acting_pid) {
            Some(s) => *s,
            None => break,
        };

        let chosen = strategy.choose_action_typed(
            &state.state,
            &state.phase,
            &acting_pid,
            plugin,
            players,
        );

        let action_type = state.phase.expected_actions[0].action_type.clone();
        let action = Action {
            action_type,
            player_id: acting_pid,
            payload: chosen,
        };
        apply_action_and_resolve_typed(plugin, &mut state, &action);
    }

    state.game_over
}

fn resolve_auto_typed<P: TypedGamePlugin>(
    plugin: &P,
    state: &mut TypedSimulationState<P::State>,
) {
    let mut max_auto = 50;
    while state.phase.auto_resolve && state.game_over.is_none() && max_auto > 0 {
        max_auto -= 1;

        let pid = if let Some(pi) = state.phase.metadata.get("player_index").and_then(|v| v.as_u64()) {
            let idx = pi as usize;
            if idx < state.players.len() {
                state.players[idx].player_id.clone()
            } else {
                "system".into()
            }
        } else {
            "system".into()
        };

        let synthetic = Action {
            action_type: state.phase.name.clone(),
            player_id: pid,
            payload: serde_json::json!({}),
        };
        apply_action_and_resolve_typed(plugin, state, &synthetic);
    }
}

fn resolve_auto(plugin: &dyn GamePlugin, state: &mut SimulationState) {
    let mut max_auto = 50;
    while state.phase.auto_resolve && state.game_over.is_none() && max_auto > 0 {
        max_auto -= 1;

        let pid = if let Some(pi) = state.phase.metadata.get("player_index").and_then(|v| v.as_u64()) {
            let idx = pi as usize;
            if idx < state.players.len() {
                state.players[idx].player_id.clone()
            } else {
                "system".into()
            }
        } else {
            "system".into()
        };

        let synthetic = Action {
            action_type: state.phase.name.clone(),
            player_id: pid,
            payload: serde_json::json!({}),
        };
        apply_action_and_resolve(plugin, state, &synthetic);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::engine::bot_strategy::{RandomStrategy, TypedRandomStrategy};
    use crate::games::carcassonne::plugin::CarcassonnePlugin;

    #[test]
    fn test_arena_random_vs_random() {
        let plugin = CarcassonnePlugin;
        let mut strategies: HashMap<String, Box<dyn BotStrategy>> = HashMap::new();
        strategies.insert("random_a".into(), Box::new(RandomStrategy));
        strategies.insert("random_b".into(), Box::new(RandomStrategy));

        let result = run_arena(
            &plugin,
            &strategies,
            3,  // just 3 games for speed
            42,
            2,
            Some(serde_json::json!({"tile_count": 10})),
            true,
            None,
        );

        assert_eq!(result.num_games, 3);
        let total_outcomes = result.wins.values().sum::<usize>() + result.draws;
        assert_eq!(total_outcomes, 3);
    }

    #[test]
    fn test_arena_typed_random_vs_random() {
        let plugin = CarcassonnePlugin;
        let mut strategies: HashMap<String, Box<dyn TypedBotStrategy<CarcassonnePlugin>>> = HashMap::new();
        strategies.insert("random_a".into(), Box::new(TypedRandomStrategy));
        strategies.insert("random_b".into(), Box::new(TypedRandomStrategy));

        let result = run_arena_typed(
            &plugin,
            &strategies,
            3,
            42,
            2,
            Some(serde_json::json!({"tile_count": 10})),
            true,
            None,
        );

        assert_eq!(result.num_games, 3);
        let total_outcomes = result.wins.values().sum::<usize>() + result.draws;
        assert_eq!(total_outcomes, 3);
    }
}
